keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1549 @@
+    // the ResourceClaim.
+    Spec ResourceClaimSpec
+    // Status describes whether the resource is available and with
+    // which attributes.
+    Status ResourceClaimStatus
+}
+
+type ResourceClaimSpec struct {
+    // ResourceClassName references the driver and additional
+    // parameters via the name of a ResourceClass that was
+    // created as part of the driver deployment.
+    //
+    // The apiserver does not check that the referenced class
+    // exists, but a driver-specific admission webhook
+    // may require that and is allowed to reject claims where
+    // the class is missing.
+    ResourceClassName string
+
+    // Parameters holds arbitrary values that will be available to the driver
+    // when allocating a resource for the claim.
+    Parameters runtime.RawExtension
+
+    // Allocation can start immediately or when a Pod wants to use
+    // the resource. Waiting for a Pod is the default.
+    AllocationMode AllocationMode
+}
+
+type AllocationMode string
+
+const (
+     AllocationModeImmediate AllocationMode = “Immediate”
+     AllocationModeDelayed AllocationMode = “Delayed”
+}
+
+type ResourceClaimStatus struct {
+   // Explains what the current status of the claim is and
+   // determines which component needs to do something.
+   Phase ResourceClaimPhase
+
+   // A copy of the driver name from the ResourceClass at
+   // the time when allocation started. Drivers can
+   // filter claims by this field. It's also necessary to
+   // support deallocation when the class gets deleted before
+   // a claim.
+   DriverName string
+
+   // When allocation is delayed, the scheduler must set
+   // the node for which it wants the resource to be allocated
+   // before the driver proceeds with allocation.
+   //
+   // For immediate allocation, the scheduler will not set
+   // this field. The resource driver controller may
+   // set it to trigger allocation on a specific node if the
+   // resources are local to nodes.
+   SelectedNode string
+
+   // When allocation is delayed, and the scheduler needs to
+   // decide on which node a Pod should run, it will first
+   // ask the driver on which nodes the resource might be
+   // made available. To trigger that check, the scheduler
+   // provides the names of nodes which might be suitable
+   // for the Pod. Will be updated periodically until
+   // the claim is allocated.
+   PotentialNodes []string
+
+   // A change of the node candidates triggers a check
+   // on which nodes the resource could be made available.
+   // Unsuitable nodes will be ignored when selecting
+   // a node. All other nodes are potential candidates,
+   // either because no information is available yet
+   // or because allocation is expected to succeed.
+   //
+   // This can change, so the driver must refresh
+   // this information periodically until a node gets
+   // selected by the scheduler.
+   UnsuitableNodes []string
```

Comment:
Is this supposed to be a subset of PotentialNodes?

Reply:
That's the intention, but it's not required.

I was already considering whether this should be a NodeSelector instead of `[]string`. If a resource driver knows exactly for which nodes it can allocate, perhaps because it isn't really consuming actual resources on a node, then the driver might also not bother with waiting for `PotentialNodes` and directly fill in this field, perhaps with a label-based selector.

Listing nodes by name would still work, it just gets a bit more complex.

Reply:
> If a resource driver knows exactly for which nodes it can allocate, perhaps because it isn't really consuming actual resources on a node

will that always be the case? Then why can't we use nodeaffinity like we do for PVs?

Reply:
No, it entirely depends on the resource driver. Some of them will have to put aside resources during allocation, others won't.

The NodeSelector part of a https://pkg.go.dev/k8s.io/api/core/v1#NodeAffinity makes sense to me. But what's the semantic of `PreferredDuringSchedulingIgnoredDuringExecution` in a `UnsuitableNodes` field?

We would have to go back to an explicit allow list before "prefer some node" makes sense. I think we need to gather some experience with a simpler v1alpha1 API before we can add "weights" to the scheduling decision.

Reply:
I was referring to the node affinity in the PV https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/persistent-volume-v1/#PersistentVolumeSpec

That and the Pod node affinity are pretty much the same. The PV only has the equivalent of `RequiredDuringSchedulingIgnoredDuringExecution`. I would expect the same for dynamic resources.

But, I want to challenge this again:

> If a resource driver knows exactly for which nodes it can allocate, perhaps because it isn't really consuming actual resources on a node

Are we assuming that all devices can be attached dynamically if the node satisfy certain criteria? What if the devices are already physically attached? What if the node runs out of those (or if a VM has an underlying limit of how many devices it can have attached to itself)?

Reply:
Updated, I hope it is clearer now.

Reply:
How does kube-scheduler know if the driver had time to update this field?

Reply:
It doesn't know that the driver had "enough" time. It will have to try periodically (because the field might be up-to-date) or when the ResourceClaim changes (which implies that new information might be available).

We could introduce some "generation" field that gets bumped by the resource driver to indicate that it has checked the ResourceClaim when no other changes were needed, but I am not sure whether periodically writing isn't more harmful than helpful.

Reply:
Initially, this field is empty. While the driver is working on filling this up, is there anything preventing scheduler from picking a candidate?

Also, in the PV/PVC life-cycle, there are no fields like `PotentialNodes` and `UnsuitableNodes`. Can you explain why is this needed here and not in PV/PVCs?

Reply:
I don't want to open a new thread on this, but this design where the scheduler and the driver go back and forth on the node selection makes me nervous (I don't know if we will be able to cover corner cases where both under up in an infinite loop of back and forth), I am wondering what other alternatives considered to avoid this completely?

Reply:
The only alternative that I see with less communication (regardless how it is done) between scheduler and resource driver is [the proposal](https://docs.google.com/document/d/1666PPUs4Lz56TqKygcy6mXkNazde-vwA7q4e5H92sUc/edit#) where the semantics of resources and availability of them is known to the kube-scheduler. Then it can "reserve" a sufficient subset of the available resources for a pod without having to consult with the third-party resource driver. Then when the pod is ready to run, those resources must be available.

That proposal got stuck because of the difficulty of describing in sufficient detail all resources that might be available and what limitations they have (network attached vs. local, how to split them up into smaller pieces, etc.).

Reply:
the open ended node lists seem like performance problems.

Reply:
I think it's worth listing such proposal in the "Alternatives" section. We have to find a balance between performance and API generality. Can you list what kind of use cases we would be missing if we were to make the scheduler more aware of resource semantics? Could we start with a subset of use cases?

BTW, on the extreme side of generality, we already have a very generic way of extending scheduler https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta2/#kubescheduler-config-k8s-io-v1beta2-Extender. I wonder if this was considered.
Although it requires the administrator to set it up and most cloud providers don't support this mechanism by default.

Reply:
> Can you list what kind of use cases we would be missing if we were to make the scheduler more aware of resource semantics?

That's exactly the problem: I don't think anyone knows a-priori which constraints the scheduler will have to be aware of for future hardware to work well in Kubernetes. It probably also will change over time, so we would end up with an API that will be forever in alpha or beta.